### PR TITLE
Don't depend on serde

### PR DESCRIPTION
**Pull request by matklad**
Okey, so I have a track record of not being able to build rls because of serde, which is understandable. I think reducing dependencies on serde may help the matters, because you won't need to juggle serde updates across several repositories. 

Moreover, rls-analysis reads the data produced by rustc_serialize, so I don't see any benefit in using serderin the first place (or is rustc switching to serde soon?).

Note that I moved `CompilerMessage` from rls here as well.

Question: why is there a lock file committed in the library package?

### CODE REVIEW

**Comment by nrc** [comment]
rustc_serialize is basically not developed any more. Serde is the preferred solution. Hopefully, the compile errors should slow down soon once the new custom derive stuff stabilises. One reason to prefer Serde, is that it will soon work on stable Rust, whereas rustc_serialize never will. It is also much faster. Finally the output of the two is not compatible, and Serde uses encodings which interact better with JS and other non-Rust users.

> Question: why is there a lock file committed in the library package?

Basically to help us build without Serde errors :-( If we have a known working setup, then with the lock file and a nightly version, we know we can compile. Without, it is (more) painful.

**Comment by matklad** [comment]
>One reason to prefer Serde, is that it will soon work on stable Rust, whereas rustc_serialize never will

Huh? rustc_serialize works perfectly fine on stable rust :) 

My main point is that serialization is an implementation detail of rls-analysis, so we can use whatever we like. And using `rustc_serialize` seems more natural just because it allows to avoid custom implementations, which are needed for serde because rustc uses rustc_serialize.

**Comment by matklad** [comment]
To be extra clear: of course we should use serde for LSP stuff and all external communications. However for parsing compiler's output `rustc_serialize` seems more natural, because that's the thing we are talking to.

**Comment by matklad** [comment]
Do you sill believe that serde is better, despite the fact that rustc_serialize **is** stable? If so, let's just close this? :)

**Comment by nrc** [comment]
> Huh? rustc_serialize works perfectly fine on stable rust :)

This surprises me, I don't see how it can possibly offer derives in stable Rust without being built into the compiler, and my understanding is that to use the version distributed with the compiler you need the `rustc_private` feature flag.

I imagine it is ok to use the plain code in stable Rust, but that is very inconvenient with the derives, which surely must be unstable.

I did originally start by using rustc_serialize rather than Serde, but had to switch. I wish I could remember the reason. It might just have been that Serde's encoding of enums was better for JS and I didn't want to use two schemes.

I do believe that Serde is where the future is - it is actively developed whereas rustc-serialize is basically dead.

**Comment by matklad** [comment]
> which surely must be unstable.

Nope :) `derive(RustcEncodable/Decodable)` works on stable since at least. Here is a link to playgroud: http://play.integer32.com/?gist=fd901e49e61f3076148f8ef6ae9aa78c&version=stable. And for example Cargo builds from stable and uses rustc_serialize for all serialization ([example](https://github.com/rust-lang/cargo/blob/c82fd7f82b8cd3a846fa63d30d6b12d79ca2abe6/src/cargo/ops/cargo_output_metadata.rs#L88-L114)).

>I wish I could remember the reason. It might just have been that Serde's encoding of enums was better for JS and I didn't want to use two schemes.

Yep, it's a bad idea to use rustc_serialize for producing JSON for external clients because the serialization is not customizable. Enums in particular are very ugly indeed. However rls-analysis types are not visible to external clients anymore. To put it in other way, we need `Decodable` for them to read compiler's output, but we don't need `Encodable` at all. 


To recap, my two points are:

* rustc_serialize is 100% stable (dead being the ultimate stable :) )

* we are reading the data produced by `rustc_serialize`, so if we use serde, we'll have to write some custom decodable implementations. Of course serde encoding is better, but rustc does not use serde.

**Comment by matklad** [comment]
Ok, so now macros 1.1 are almost here, I think this can be closed: performance benefits of serde outweigh the need to write extra code :)
